src/Http/Routing/src/Matching/DfaMatcherBuilder.cs

```diff
@@ -264,10 +265,52 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
                                 };
                             }
 
-                            // A parameter should traverse all literal nodes as well as the parameter node
                             if (parent.Literals != null)
                             {
-                                nextParents.AddRange(parent.Literals.Values);
+                                // If the parameter contains constraints, we can be smarter about it and evaluate them while we build the tree.
+                                // If the literal doesn't match any of the constraints, we can prune the branch.
+                                // For example, for a parameter in a route {lang:length(2)} and a parent literal "ABC", we can check that "ABC"
+                                // doesn't meet the parameter constraint (length(2)) when building the tree, and avoid the extra nodes.
+                                if (endpoint.RoutePattern.ParameterPolicies.TryGetValue(parameterPart.Name, out var parameterPolicyReferences))
+                                {
+                                    // The list of parameters that fail to meet at least one ILiteralConstraint.
+                                    var hasFailingPolicy = new bool[parent.Literals.Keys.Count];
+
+                                    // Whether or not all parameters have failed to meet at least one constraint.
+                                    for (var k = 0; k < parameterPolicyReferences.Count; k++)
+                                    {
+                                        var reference = parameterPolicyReferences[k];
+                                        var parameterPolicy = _parameterPolicyFactory.Create(parameterPart, reference);
```

Comment:
When do constraints normally get created and where are they cached?

Reply:
It's cached in the reference https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/ParameterPolicyFactory.cs#L37